src/client.js

```diff
@@ -3,36 +3,43 @@
  */
 import 'babel/polyfill';
 import React from 'react';
-import BrowserHistory from 'react-router/lib/BrowserHistory';
-import Location from 'react-router/lib/Location';
-import queryString from 'query-string';
+import ReactDOM from 'react-dom';
+import createHistory from 'history/lib/createBrowserHistory';
+import createLocation from 'history/lib/createLocation';
 import createStore from './redux/create';
 import ApiClient from './helpers/ApiClient';
 import universalRouter from './helpers/universalRouter';
-const history = new BrowserHistory();
+const history = createHistory();
+
 const client = new ApiClient();
 
 const dest = document.getElementById('content');
 const store = createStore(client, window.__data);
-const search = document.location.search;
-const query = search && queryString.parse(search);
-const location = new Location(document.location.pathname, query);
-universalRouter(location, history, store)
-  .then(({component}) => {
-    React.render(component, dest);
-    if (__DEVTOOLS__) {
-      const { DevTools, DebugPanel, LogMonitor } = require('redux-devtools/lib/react');
-      React.render(<div>
-        {component}
-        <DebugPanel top right bottom key="debugPanel">
-          <DevTools store={store} monitor={LogMonitor}/>
-        </DebugPanel>
-      </div>, dest);
-    }
-  }, (error) => {
-    console.error(error);
-  });
+const location = createLocation(document.location.pathname, document.location.search);
+
+const render = (loc, hist, str, preload) => {
+  universalRouter(loc, hist, str, preload)
+    .then(({component}) => {
+      ReactDOM.render(component, dest);
+      if (__DEVTOOLS__) {
+        const { DevTools, DebugPanel, LogMonitor } = require('redux-devtools/lib/react');
+        ReactDOM.render(<div>
+          {component}
+          <DebugPanel top right bottom key="debugPanel">
+            <DevTools store={store} monitor={LogMonitor}/>
+          </DebugPanel>
+        </div>, dest);
+      }
+    }, (error) => {
+      console.error(error);
+    });
+};
+
+history.listen((loc) => {
+  render(loc, history, store, true);
```

Comment:
I just pulled it down and tested it--it seems to be updating the location state before the request completes, which I believe is not how it should operate.

I was looking into Lifecycle hooks to prevent transition so that we could accomplish that: https://github.com/rackt/react-router/issues/2061